ocaml/xapi/extauth_plugin_ADwinbind.ml

```diff
@@ -31,6 +45,245 @@ let tdb_tool = !Xapi_globs.tdb_tool
 
 let debug_level = !Xapi_globs.winbind_debug_level |> string_of_int
 
+let ntlm_auth uname passwd : (unit, exn) result =
+  try
+    let args = ["--username"; uname] in
+    let _stdout =
+      Helpers.call_script ~log_output:Never ~stdin:passwd
+        !Xapi_globs.ntlm_auth_cmd args
+    in
+    Ok ()
+  with e -> Error e
+
+let call_ldap (query : string) : (string, exn) result =
+  try
+    let args = ["ads"; "search"; "--machine-pass"; query] in
+    let stdout =
+      Helpers.call_script ~log_output:On_failure !Xapi_globs.net_cmd args
+    in
+    Ok stdout
+  with e -> Error e
+
+let call_wbinfo (args : string list) : (string, exn) result =
+  try
+    (* we trust wbinfo will not print any sensitive info on failure *)
+    let stdout = Helpers.call_script ~log_output:On_failure wb_cmd args in
+    Ok stdout
+  with e -> Error e
+
+module Ldap = struct
+  type user = {
+      upn: string
+    ; account_disabled: bool
+    ; account_expired: bool
+    ; account_locked: bool
+    ; password_expired: bool
+  }
+  [@@deriving rpcty]
+
+  let string_of_user x =
+    Rpcmarshal.marshal user.Rpc.Types.ty x |> Jsonrpc.to_string
+
+  let parse_user stdout : (user, string) result =
+    let module Map = Map.Make (String) in
+    let module P = struct
+      open Angstrom
+
+      let is_space = function ' ' -> true | _ -> false
+
+      let s = take_while is_space
```

Comment:
While I think `ws` is acceptable as a name in a scanner, this is a step too far and would prefer a longer name. I understand the problem of having similar names for predicates and parsers which you solved by using `is_` as prefix. I think this should be `spaces`. I am a bit surprised we use both spaces and white space - this could be confusing for the user.

Reply:
I've doing some renaming :+1: 

`s` was used for separating key values pairs like `key: value`, but I was probably being too generous accepting multiple spaces, so I've changed it to just `char ' '` instead